### DISCUSSION TITLE
Redirect based on users ip

### DISCUSSION THREAD

**Discussion by TomGriffiths88** [vercel/next.js#17713] [Help] [answered]
I've got a project I've built out in next js and the client has just told me that we can't have the site viewable by visitors from certain locations, due to no compete agreement. 

Is there a way I can put a hard redirect from the host 'Vercel' so anyone making a request in say 'france' will be redirected to the French site? 

Or do I need to send a request from server side to an IP lookup API and then feed that into a context on the client? 

Ideally looking for the most simple solution. I would like to keep all SSG pages for optimal speed if possible. 

Many thanks in advance

**Comment by lfades** [comment, marked as answer]
We'll be doing something like this for the incoming built-in [i18n](https://github.com/vercel/next.js/discussions/17078) support.

For now, you can create an API route that checks for the IP and redirects based on that

**Reply by TomGriffiths88** [reply]
Thanks man. I will do just that.

**Reply by julienben** [reply]
Hey there, Trying to understand how you set this up as I have similar requirements.

Do you run a GeoIP request in `getServerSideProps`? Is it possible to do it in a way that doesn't hurt static generation?

I need to redirect the user to a different page based on the result of the GeoIP request but the props passed to each page are independent of that...

**Reply by TomGriffiths88** [reply]
What I did in the end. Was to create a context for the user location. Within the context provider function I sent a request to an API route which hits an IP checking service (IP2location) and returns the results to the context. If the location is on the blacklist, I just redirect the user to the another site via the router but if it's all good I save it to the context state so I can then access this anywhere in the app. 

This way all the pages are still static, and I just show a spinner whilst the context grabs ip information on first render. Very quick and super smooth. 

If your page props depend on the IP, then maybe load a basic template or skeleton statically and fetch the data in a useEffect hook once you receive the context IP?

**Reply by julienben** [reply]
Thanks for your quick and detailed response!!

Sounds like what you implemented is very similar to [this article](https://www.sitepoint.com/ip2location-next-js-divert-traffic/)'s suggestion.

What I'm being asked to build is closer to what for example Microsoft.com does. When you try to access the root URL, you get a 301 redirect to the relevant version (i.e. /en-us or /fr-fr). I need to decide which page to send the user to according to their IP and a Cookie. I just wanted to know if I could somehow make this request in a way that doesn't interfere with static generation but I have a feeling that's not gonna be possible...

**Reply by TomGriffiths88** [reply]
Thats basically what I did but changed it so it would work with static. 

Sounds like you will be able to use the new i18n in nextJS 10. Its out now [Next JS i18n](https://nextjs.org/docs/advanced-features/i18n-routing).

Much simpler to setup. And works great.

**Reply by julienben** [reply]
FYI, I did end up using Next's i18n feature but it doesn't take care of the IP redirect side of things. I need it to be a server side redirect, not client side.

I was really worried that I would need to do it in `getServerSideProps` and lose all the benefits of static generation but I stumbled into the perfect solution today. A Cloudflare Worker which sits between the browser and the Next.js server. It works perfectly well and has essentially no impact on performance.

https://developers.cloudflare.com/workers/examples/country-code-redirect

**Reply by KalanaPerera** [reply]
@julienben  Thank you for the amazing solution
